### ISSUE TITLE
Saving & Loading ShapeletModel / SerializableShapeletModel

### ISSUE THREAD

**Issue by slkpy** [tslearn-team/tslearn#178]
Hi,

First, thanks for this work, I'm really glad that I could find and use properly the Shapelets.

I know there are currently several PR related to saving Shapelets, especially the huge WIP #120.
What I'd like to know is whether it's currently possible to train a ShapeletModel, save its relevant parameters so that they can used to "rebuild" the ShapeletModel in another python instance and used to locate Shapelets on other time series.

I've searched through the tslearn GitHub repository the past few days and here's what I found :
- gio8tisu created the  PR #53  for this very issue in July 2018 and has proposed save() and load_model() methods, but as I understand it, you can't use locate() with a mock model you've filled with .load_model().
- there's a class "SerializableShapeletModel" based on "ShapeletModel" (though the accuracy seems rather bad compared to ShapeletModel), that can be dumped, but when I try to load it, I get the same error as mentionned in PR #85 

Please enlighten me !

**Comment by johannfaouzi** [comment]
MWE (very ugly but it works)

```python
from joblib import dump, load
from keras.models import load_model
from tslearn.shapelets import ShapeletModel, LocalSquaredDistanceLayer, GlobalMinPooling1D
from tslearn.datasets import UCR_UEA_datasets


# Load a dataset
X_train, y_train, X_test, y_test = UCR_UEA_datasets().load_dataset('Coffee')

# Define and fit an instance of ShapeletModel
clf = ShapeletModel(random_state=42)
clf.fit(X_train, y_train)

# Save the model by accessing the 'model_' attribute
clf.model_.save('shapelet_model.h5')

# Save all the attributes
clf.model_ = None
clf.transformer_model_ = None
clf.locator_model_ = None
dump(clf.__dict__, 'attributes.gz')

# Delete the instance
del clf

# Create a new instance
clf = ShapeletModel()

# Define the 'model_' attribute of this instance
model = load_model(
    'shapelet_model.h5',
    custom_objects={'LocalSquaredDistanceLayer': LocalSquaredDistanceLayer,
                    'GlobalMinPooling1D': GlobalMinPooling1D}
)
clf.model_ = model

# Define the other necessary attributes
attributes = load('attributes.gz')
clf._X_fit = attributes['_X_fit']
clf.d_ = attributes['d_']
clf.classes_ = attributes['classes_']
clf.categorical_y_ = attributes['categorical_y_']
clf.label_binarizer_ = attributes['label_binarizer_']

# Predict on the test set
clf.score(X_test, y_test)
```

Adding `save` and `load` methods could be useful...

**Comment by rtavenar** [comment]
Thanks @johannfaouzi !

I guess the load and save methods could be implemented in the scope of PR #160 (base attributes could be loaded/saved using standard means and the model itself could be dealt with through custom code, WDYT?

**Comment by slkpy** [comment]
Hi @johannfaouzi ,

Thanks for your code sample. However, I does not solve my problem and I still have the following issues :  

1. First, my usage of Shapelets is to locate a known signature on time series, hence I need to use the .locate() method on the reloaded model, but when I try to do so, I have a traceback stating there's no 'locator_model_' attribute in my ShapeletModel, which makes sense since it's been set to None before dumping the attributes.
If I don't set the locator_model_ to None before dumping the attributes, I will have a traceback when trying to load the model attributes.

2. Also, by defining the ShapeletModel the way you did _(i.e. "clf = ShapeletModel(random_state=42)")_, I don't get any shapelet once its fitted.
On the other hand, defining the ShapeletModel the "traditional way" (see [1] below), I have the same traceback when I try to dump the model dictionary that I have when I try to pickle the whole ShapeletModel, probably because the dictionary it contains the "optimizer" attribute.

[1] Here's what I mean by "traditional way" : 

> clf = ShapeletModel(n_shapelets_per_size=shapelet_sizes,
>                         optimizer=Adagrad(lr=.1),
>                         weight_regularizer=.01,
>                         max_iter=50,
>                         verbose=0)

Therefore, any additional idea is most welcome :)

@rtavenar my understanding was that PR #160 would feature methods to save and load the whole ShapeletModel (and get rid of the SerialisableShapeletModel class), and I would very much prefer this since I really need to use the model locator.

Thanks !

**Comment by johannfaouzi** [comment]
You just need to do the same for the other keras-related attributes (`locator_model_` and `transformer_model_`) and for the `n_shapelets_per_size` dictionary:

```python
from joblib import dump, load
from keras.models import load_model
from tslearn.shapelets import (
    ShapeletModel, LocalSquaredDistanceLayer, GlobalMinPooling1D, GlobalArgminPooling1D
)
from tslearn.datasets import UCR_UEA_datasets


# Load a dataset
X_train, y_train, X_test, y_test = UCR_UEA_datasets().load_dataset('Coffee')

# Define and fit an instance of ShapeletModel
clf = ShapeletModel(n_shapelets_per_size={12: 30, 24: 30}, random_state=42)
clf.fit(X_train, y_train)

# Save all the keras-related attributes by accessing the 'model_' attribute
clf.model_.save('model.h5')
clf.transformer_model_.save('transformer_model.h5')
clf.locator_model_.save('locator_model.h5')

# Save all the attributes
clf.model_ = None
clf.transformer_model_ = None
clf.locator_model_ = None
dump(clf.__dict__, 'attributes.gz')

# Delete the instance
del clf

# Create a new instance
clf = ShapeletModel()

# Define the 'model_' attribute of this instance
model = load_model(
    'model.h5',
    custom_objects={'LocalSquaredDistanceLayer': LocalSquaredDistanceLayer,
                    'GlobalMinPooling1D': GlobalMinPooling1D}
)
transformer_model = load_model(
    'transformer_model.h5',
    custom_objects={'LocalSquaredDistanceLayer': LocalSquaredDistanceLayer,
                    'GlobalMinPooling1D': GlobalMinPooling1D}
)
locator_model = load_model(
    'locator_model.h5',
    custom_objects={'LocalSquaredDistanceLayer': LocalSquaredDistanceLayer,
                    'GlobalMinPooling1D': GlobalMinPooling1D,
                    'GlobalArgminPooling1D': GlobalArgminPooling1D}
)

clf.model_ = model
clf.transformer_model_ = transformer_model
clf.locator_model_ = locator_model

# Define the other necessary attributes
attributes = load('attributes.gz')
clf.n_shapelets_per_size = attributes['n_shapelets_per_size']
clf._X_fit = attributes['_X_fit']
clf.d_ = attributes['d_']
clf.classes_ = attributes['classes_']
clf.categorical_y_ = attributes['categorical_y_']
clf.label_binarizer_ = attributes['label_binarizer_']

# Locate some time series
clf.locate(X_test)

# Get learned shapelets
clf.shapelets_
```

**Comment by slkpy** [comment]
Thanks @johannfaouzi , it does work this way !

However, since the results were bad with the ShapeletModel defined the way you did, I tried changing the ShapeletModel definition with the one provided in the "Learning Shapelets" example that gave good results, that is : 

_shp_clf = ShapeletModel(n_shapelets_per_size=shapelet_sizes,
                        optimizer=Adagrad(lr=.1),
                        weight_regularizer=.01,
                        max_iter=50,
                        verbose=0)_

Unfortunately, by doing so, I get a traceback when I get to _dump(clf.__dict__, 'attributes.gz')_

After investigating, I realized the problem was caused by defining the optimizer as a parameter (even though I don't understand why). Hence, the following ShapeletModel definition worked fine and enabled me to save and load the ShapeletModel afterwards : 

_shp_clf = ShapeletModel(n_shapelets_per_size=shapelet_sizes,
                        weight_regularizer=.01,
                        max_iter=50,
                        verbose=0)_

As I still got bad results with the default optimizer (SGD) and learning rate, I changed those default values in the shapelets.py and optimizers.py in my virtual environment so that the ShapeletModel definition will by default use the Adagrad optimizer and a learning rate of 0.1. And it did work just fine !

Now I realize this is even dirtier but at this point I don't even know where to look (tslearn or keras) to understand the problem. If you have any tip, it's welcomed, but in any case **this dirty method does work**.
